### ISSUE TITLE
Support Custom Datastore

### ISSUE THREAD

**Issue by dominikschlosser** [keycloak/keycloak#25830]
### Description

Since `MapStorageSpi` and `MAP_STORAGE`-feature have been removed, it became quite challenging to replace Keycloak storage including Infinispan-caching with a custom implementation. I understand that supporting something like map store was to big in scope and very challenging for the core team. Thus i am proposing a kind of middle ground: A CustomDatastore-feature which disables all of the default storage providers to simplify replacing them with custom implementations.

If you support this idea, i could provide the implementation.



### Discussion

https://github.com/keycloak/keycloak/discussions/24052

### Motivation

Right now, to replace default datastore one not only has to implement `DatastoreProvider` but also needs to make sure that default implementations are not loaded (as they register themselves as event listeners, start infinispan-caches, do JPA-database migrations, etc inside of `init()`).
This was taken care of in KC < 23 when `Feature.MAP_STORAGE` was enabled, as all "Legacy"-providers where then disabled via `EnvironmentDependentProviderFactory.isSupported()`.

This is no longer the case in KC 23. Thus to achieve the same we did not find another way than either disabling all the default providers (cumbersome) or overriding them by using the same ID and a higher `order()` (weird).
And currently there is no way at all to disable JPA/Cache Quarkus-BuildSteps.

### Details

- Add a `CUSTOM_DATASTORE`-feature (experimental)
- If this feature is activated, all default storage providers are deactivated via `EnvironmentDependentProviderFactory` (basically replace all occurences of `!Profile.isFeatureEnabled(Profile.Feature.MAP_STORAGE)` from KC < 23 with `!Profile.isFeatureEnabled(Profile.Feature.CUSTOM_DATASTORE)`
- Introduce `IsCustomDatastoreEnabled` BooleanSupplier (based on `IsLegacyStoreEnabled` from KC < 23) to disable Cache- and JPA-Quarkus-BuildSteps

**Comment by sschu** [comment]
@ahus1 FYI.

**Comment by opdt** [comment]
@mhajas FYI, this proposal is the result of our effort to replace Datastore and Caching layer with Cassandra in Keycloak 23.

**Comment by mhajas** [comment]
Thank you for this suggestion. I find `CUSTOM_DATASTORE` feature a bit too generic and hard to understand what exactly it means. Do you think the name could be more self-explaining? For example, if you need to disable caching we can have a feature `DISABLED_CACHING`.  And the same for other changes you need. WDYT?

**Comment by dominikschlosser** [comment]
Ok we would need `DISABLED_CACHING` and `DISABLED_JPA_DATASTORE` to disable schema migrations etc.


**Comment by dominikschlosser** [comment]
You are right but it would be nice to be able to deactivate schema migrations and the JPA Quarkus-BuildStep.

**Comment by mhajas** [comment]
@dominikschlosser I was thinking about this and I am not sure I fully understand all the problems you are facing.

So in a nutshell, you have a new implementation of the storage layer and to use it you need to disable some providers. Currently, you are disabling providers using configuration options and it works? However, you don't like this approach and you would like to replace it with a feature so you can replace many config options (disabling providers you don't need) with a few options (enabling features that disable the providers in the background).

The only outstanding things, that cannot be solved by configuration options at the moment, are the caching and JPA Quarkus build steps that are running all the time. This is a problem that should be solved but the feature is not the correct solution. The correct solution is to run these only when JPA/Infinispan connection providers are enabled. Using a feature for controlling them would mean we are assuming that JPA/Infinispan connection providers can be disabled only using the feature but they can still be disabled with the config options.

Adding a new feature sounds a little bit like abusing how features work. A feature represents a new functionality that is added to Keycloak and by enabling it you can enable the functionality and consume what it provides. While here it is a grouping for a bunch of configuration options.

On the other hand, what you are doing is legitimate usage of Keycloak and we want to have our storage layer flexible enough to allow you to use it the way you want. We need to remove any obstacles you are having (like the build steps that are running when they should not) so if there is anything else you are facing please raise it here.

**Comment by dominikschlosser** [comment]
> So in a nutshell, you have a new implementation of the storage layer and to use it you need to disable some providers. Currently, you are disabling providers using configuration options and it works? However, you don't like this approach and you would like to replace it with a feature so you can replace many config options (disabling providers you don't need) with a few options (enabling features that disable the providers in the background).

Correct, but currently we override certain providers using the same ID as the default ones to prevent the default providers from being initialized. This is very hacky and wasn't necessary pre-Keycloak 23 since Feature.MAP_DATASTORE disabled all the "Legacy providers". This is were the "Feature.CUSTOM_DATASTORE / DISABLE_CACHING" idea came from.

> The only outstanding things, that cannot be solved by configuration options at the moment, are the caching and JPA Quarkus build steps that are running all the time. This is a problem that should be solved but the feature is not the correct solution. The correct solution is to run these only when JPA/Infinispan connection providers are enabled. Using a feature for controlling them would mean we are assuming that JPA/Infinispan connection providers can be disabled only using the feature but they can still be disabled with the config options.

This seems like a good solution to me. 

> Adding a new feature sounds a little bit like abusing how features work. A feature represents a new functionality that is added to Keycloak and by enabling it you can enable the functionality and consume what it provides. While here it is a grouping for a bunch of configuration options.

Following this argument, the feature should be "Feature.INFINISPAN_CACHING" which would be enabled by default and disabling it would disable all the default Infinispan-providers, right?

**Comment by dominikschlosser** [comment]
> Providers should not be initialized if they are disabled by the enabled flag.
Yes but there are a lot of providers to disable. It was mainly a ease-of-use-decision to use `order()`.

> Also, when you specify a provider ID in the configuration it should not initialize any other provider for that SPI.
Yep but same reason, would be nice if we just had to override Datastore provider which transitively defines a bunch of other SPI implementations.

**Comment by mhajas** [comment]
> Yes but there are a lot of providers to disable. It was mainly a ease-of-use-decision to use `order()`.

> Yep but same reason, would be nice if we just had to override Datastore provider which transitively defines a bunch of other SPI implementations.

But that is what the configuration is for. To me, it makes sense to set a lot of configuration options when many providers need to be disabled/replaced. I don't see this as a reason to introduce a new feature. But I see it as a potential follow-up in the future if your concept works well.

We can gather more feedback on this. Can you please start a new discussion dedicated to this topic? The currently linked issue `Discontinuation of Map Store: Implications` is too generic.

btw. I created an enhancement issue to not run the persistence-related Quarkus build steps when the JPA connection provider is disabled: https://github.com/keycloak/keycloak/issues/26054

**Comment by dominikschlosser** [comment]
I summarized our problems and solution proposal here for discussion: https://github.com/keycloak/keycloak/discussions/26077

**Comment by dominikschlosser** [comment]
With `STATELESS` this is not needed anymore.

**Comment by ahus1** [comment]
We would love to have feedback on the new stateless feature. See https://github.com/keycloak/keycloak/issues/50642 for the progress to make it supported, and link to the discussion.
